tests/integration/test_charm.py

```diff
@@ -0,0 +1,122 @@
+SLURMDBD = "slurmdbd"
+SLURMCTLD = "slurmctld"
+DATABASE = "percona-cluster"
+UNIT = f"{SLURMDBD}/0"
+
+
+@pytest.mark.abort_on_fail
+@pytest.mark.skip_if_deployed
+@pytest.mark.parametrize("series", SERIES)
+async def test_build_and_deploy_against_edge(
+    ops_test: OpsTest, slurmdbd_charm: Coroutine[Any, Any, pathlib.Path], series: str
+) -> None:
+    """Test that the slurmdbd charm can stabilize against slurmctld and percona."""
+    logger.info(f"Deploying {SLURMDBD} against {SLURMCTLD} and {DATABASE}...")
+    slurmctld_res = get_slurmctld_res()
+    get_slurmdbd_res()
+    await asyncio.gather(
+        ops_test.model.deploy(
+            str(await slurmdbd_charm),
+            application_name=SLURMDBD,
+            num_units=1,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            SLURMCTLD,
+            application_name=SLURMCTLD,
+            channel="edge",
+            num_units=1,
+            resources=slurmctld_res,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            DATABASE,
+            application_name=DATABASE,
+            channel="edge",
+            num_units=1,
+            series="bionic",
+        ),
+    )
+    # Attach resources to charms.
+    await ops_test.juju("attach-resource", SLURMCTLD, f"etcd={slurmctld_res['etcd']}")
+    # Set integrations for charmed applications.
+    await ops_test.model.relate(f"{SLURMDBD}:{SLURMDBD}", f"{SLURMCTLD}:{SLURMDBD}")
+    await ops_test.model.relate(f"{SLURMDBD}:db", f"{DATABASE}:db")
+    # Reduce the update status frequency to accelerate the triggering of deferred events.
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(apps=[SLURMDBD], status="active", timeout=1000)
+        assert ops_test.model.applications[SLURMDBD].units[0].workload_status == "active"
+
+
+@pytest.mark.abort_on_fail
+@tenacity.retry(
+    wait=tenacity.wait.wait_exponential(multiplier=2, min=1, max=30),
+    stop=tenacity.stop_after_attempt(3),
+    reraise=True,
+)
+async def test_slurmdbd_is_active(ops_test: OpsTest) -> None:
+    """Test that slurmdbd is active inside Juju unit."""
+    logger.info("Checking that slurmdbd daemon is active inside unit...")
+    async with unit_connection(ops_test, SLURMDBD, UNIT) as conn:
+        stdin, stdout, stderr = conn.exec_command("systemctl is-active slurmdbd")
+        assert codecs.decode(stdout.read()).strip("\n") == "active"
+
+
+@pytest.mark.abort_on_fail
+@tenacity.retry(
+    wait=tenacity.wait.wait_exponential(multiplier=2, min=1, max=30),
+    stop=tenacity.stop_after_attempt(3),
+    reraise=True,
+)
+async def test_slurmdbd_port_listen(ops_test: OpsTest) -> None:
+    """Test that slurmdbd is listening on port 6819."""
+    logger.info("Checking that slurmdbd is listening on port 6819...")
+    async with unit_connection(ops_test, SLURMDBD, UNIT) as conn:
+        stdin, stdout, stderr = conn.exec_command("sudo lsof -i -n | grep ':6819'")
```

Comment:
The pipe of the `lsof` into `grep` is unnecessary. If you check the man page of lsof, you'll find that you can get the PID of the processes that using TCP network files on port 5433 in state of LISTEN. Try the following:

```bash
sudo lsof -t -n -iTCP:5433 -sTCP:LISTEN
```

Reply:
I mostly just piped it to `grep` since it was quick and easy (and I was more familiar with grepping compared to "list open files"). I will look into using `lsof` with the extra flags.

Reply:
Resolved in https://github.com/canonical/slurmdbd-operator/pull/4/commits/05ac3889ef35a7fc621427d571247016b43f8dca